src/test/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImplTest.java

```diff
@@ -0,0 +1,311 @@
+import org.jenkinsci.plugins.workflow.cps.CpsThread;
+import org.jenkinsci.plugins.workflow.cps.nodes.DescriptorMatchPredicate;
+import org.jenkinsci.plugins.workflow.cps.nodes.StepStartNode;
+import org.jenkinsci.plugins.workflow.flow.FlowExecution;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.graphanalysis.DepthFirstScanner;
+import org.jenkinsci.plugins.workflow.graphanalysis.FlowScanningUtils;
+import org.jenkinsci.plugins.workflow.graphanalysis.ForkScanner;
+import org.jenkinsci.plugins.workflow.graphanalysis.LinearScanner;
+import org.jenkinsci.plugins.workflow.graphanalysis.NodeStepTypePredicate;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.jenkinsci.plugins.workflow.steps.EchoStep;
+import org.jenkinsci.plugins.workflow.support.storage.SimpleXStreamFlowNodeStorage;
+import org.jenkinsci.plugins.workflow.test.steps.SemaphoreStep;
+import org.junit.Assert;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.BuildWatcher;
+import org.jvnet.hudson.test.JenkinsRule;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Tests the input sanitization and step persistence here
+ */
+public class ArgumentsActionImplTest {
+    // Run pipeline, with pauses for steps
+    @ClassRule
+    public static BuildWatcher buildWatcher = new BuildWatcher();
+
+    @Rule
+    public JenkinsRule r = new JenkinsRule();
+
+    /** Helper function to test direct file deserialization for an execution */
+    private void testDeserialize(FlowExecution execution) throws Exception {
+        if (!(execution instanceof CpsFlowExecution) || !(((CpsFlowExecution)execution).getStorage() instanceof SimpleXStreamFlowNodeStorage)) {
+            return;  // Test is unfortunately coupled to the implementation -- otherwise it will simply hit caches
+        }
+
+        SimpleXStreamFlowNodeStorage storage = (SimpleXStreamFlowNodeStorage)(((CpsFlowExecution)execution).getStorage());
+        Method getFileM = SimpleXStreamFlowNodeStorage.class.getDeclaredMethod("getNodeFile", String.class);
+        getFileM.setAccessible(true);
+
+        List<FlowNode> nodes = new DepthFirstScanner().allNodes(execution.getCurrentHeads());
+        Collections.sort(nodes, FlowScanningUtils.ID_ORDER_COMPARATOR);
+
+        Field nodeExecutionF = FlowNode.class.getDeclaredField("exec");
+        nodeExecutionF.setAccessible(true);
+
+        // Read each node via deserialization from storage, and sanity check the node, the actions, and the ArgumentsAction read back right
+        for (FlowNode f : nodes) {
+            XmlFile file = (XmlFile)(getFileM.invoke(storage, f.getId()));
```

Comment:
@jglick Yes, this is covering a REAL class of significant issue I saw during testing -- serialization linked problems.  I want to ensure we have some coverage for this because the ArgumentsAction exposes more of this. 

 As proven by the bug you called my attention to this very afternoon, I was correct.  ;-)